File: app/src/test/java/com/mohsenoid/rickandmorty/util/config/ConfigProviderTest.java
package com.mohsenoid.rickandmorty.util.config;

import android.app.Application;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.annotation.Implements;
import org.robolectric.shadows.ShadowConnectivityManager;
import org.robolectric.shadows.ShadowNetworkInfo;

import static androidx.test.core.app.ApplicationProvider.getApplicationContext;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.robolectric.Shadows.shadowOf;

@Config(sdk = Build.VERSION_CODES.P)
@RunWith(RobolectricTestRunner.class)
@Implements(ShadowConnectivityManager.class)
public class ConfigProviderTest {

    private ShadowNetworkInfo shadowOfActiveNetworkInfo;

    private ConfigProvider configProvider;

    @Before
    public void setUp() {
        Application application = getApplicationContext();

        ConnectivityManager connectivityManager = (ConnectivityManager) application.getSystemService(Context.CONNECTIVITY_SERVICE);
        assert connectivityManager != null;
        shadowOfActiveNetworkInfo = shadowOf(connectivityManager.getActiveNetworkInfo());

        configProvider = ConfigProviderImpl.getInstance(application);
    }

    @After
    public void tearDown() {
        ConfigProviderImpl.instance = null;
    }

    @Test
    public void testIsOnline() {
        // GIVEN
        shadowOfActiveNetworkInfo.setConnectionStatus(NetworkInfo.State.CONNECTED);

        // WHEN
        boolean isOnline = configProvider.isOnline();

        //THEN
        assertTrue(isOnline);
    }

    @Test
    public void testIsOffline() {
        // GIVEN
        shadowOfActiveNetworkInfo.setConnectionStatus(NetworkInfo.State.DISCONNECTED);

        // WHEN
        boolean isOnline = configProvider.isOnline();

        //THEN
        assertFalse(isOnline);
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/util/config/ConfigProviderTest.kt
package com.mohsenoid.rickandmorty.util.config

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.os.Build
import androidx.test.core.app.ApplicationProvider
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.koin.core.context.stopKoin
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import org.robolectric.annotation.Config
import org.robolectric.annotation.Implements
import org.robolectric.shadows.ShadowConnectivityManager
import org.robolectric.shadows.ShadowNetworkInfo

@Config(sdk = [Build.VERSION_CODES.LOLLIPOP_MR1])
@RunWith(RobolectricTestRunner::class)
@Implements(ShadowConnectivityManager::class)
class ConfigProviderTest {

    private lateinit var shadowOfActiveNetworkInfo: ShadowNetworkInfo

    private lateinit var configProvider: ConfigProvider

    @Before
    fun setUp() {
        val application: Application = ApplicationProvider.getApplicationContext()
        val connectivityManager: ConnectivityManager =
            application.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        shadowOfActiveNetworkInfo = Shadows.shadowOf(connectivityManager.activeNetworkInfo)
        configProvider = ConfigProviderImpl(application)
    }

    @After
    fun tearDown() {
        stopKoin()
    }

    @Test
    fun `test isOnline`() {
        // GIVEN
        @Suppress("DEPRECATION")
        shadowOfActiveNetworkInfo.setConnectionStatus(NetworkInfo.State.CONNECTED)

        // WHEN
        val isOnline: Boolean = configProvider.isOnline()

        // THEN
        isOnline.shouldBeTrue()
    }

    @Test
    fun `test isOffline`() {
        // GIVEN
        @Suppress("DEPRECATION")
        shadowOfActiveNetworkInfo.setConnectionStatus(NetworkInfo.State.DISCONNECTED)

        // WHEN
        val isOnline: Boolean = configProvider.isOnline()

        // THEN
        isOnline.shouldBeFalse()
    }
}
